### PR TITLE
OHRM5X-1655: Develop dashboard - employees on leave API

### PR DESCRIPTION
**Pull request by ManulMax**

### CODE REVIEW

File: src/plugins/orangehrmLeavePlugin/Dao/LeaveListDao.php [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+        $this->setSortingAndPaginationParams($q, $leaveListSearchFilterParams);
+
+        $q->andWhere('leaveList.date = :date')->setParameter('date', $leaveListSearchFilterParams->getDate());
+        $q->andWhere($q->expr()->orX(
+            $q->expr()->eq('leaveList.status', ':pending'),
+            $q->expr()->eq('leaveList.status', ':approved')
+        ))
+        ->setParameter('pending', Leave::LEAVE_STATUS_LEAVE_PENDING_APPROVAL)
+        ->setParameter('approved', Leave::LEAVE_STATUS_LEAVE_APPROVED);
```
Comment (RajithaKumara) [inline src/plugins/orangehrmLeavePlugin/Dao/LeaveListDao.php:59]: `Leave::LEAVE_STATUS_LEAVE_TAKEN` also needs to consider right. I saw in spec

File: src/plugins/orangehrmLeavePlugin/Service/LeaveListService.php [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+namespace OrangeHRM\Leave\Service;
+
+use OrangeHRM\Leave\Dao\LeaveListDao;
+
+class LeaveListService
+{
+    /**
+     * @var LeaveListDao|null
+     */
+    private ?LeaveListDao $leaveListDao = null;
```
Comment (RajithaKumara) [inline src/plugins/orangehrmLeavePlugin/Service/LeaveListService.php:29]:
```suggestion
    /**
     * @var LeaveListDao
     */
    private LeaveListDao $leaveListDao;
```
:bulb:

File: src/plugins/orangehrmLeavePlugin/Service/LeaveListService.php [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+    /**
+     * @return LeaveListDao
+     */
+    public function getLeaveListDao(): LeaveListDao
+    {
+        if (!$this->leaveListDao instanceof LeaveListDao) {
+            $this->leaveListDao = new LeaveListDao();
+        }
+        return $this->leaveListDao;
```
Comment (RajithaKumara) [inline src/plugins/orangehrmLeavePlugin/Service/LeaveListService.php:39]:
```suggestion
        return $this->leaveListDao ??= new LeaveListDao();
```
:bulb: 
E.g.
https://github.com/orangehrm/orangehrm/blob/5.2-ldap/src/plugins/orangehrmLDAPAuthenticationPlugin/Auth/LDAPAuthProvider.php#L28-L36

File: src/plugins/orangehrmLeavePlugin/Service/Model/LeaveListModal.php [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+        $attributeNames = [
+            'id',
+            'date',
+            'lengthHours',
+            ['employee','empNumber'],
+            ['employee','lastName'],
+            ['employee','firstName'],
+            ['employee','middleName'],
+            ['employee','employeeId'],
```
Comment (RajithaKumara) [inline src/plugins/orangehrmLeavePlugin/Service/Model/LeaveListModal.php:59]: :ok_hand:

File: src/plugins/orangehrmLeavePlugin/Service/Model/LeaveListModal.php [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        $this->setEntity($leave);
+        $filter = [
+            'id',
+            ['getDecorator', 'getLeaveDate'],
+            'lengthHours',
+            ['getEmployee','getEmpNumber'],
+            ['getEmployee','getLastName'],
+            ['getEmployee','getFirstName'],
+            ['getEmployee','getMiddleName'],
+            ['getEmployee','getEmployeeId'],
```
Comment (RajithaKumara) [inline src/plugins/orangehrmLeavePlugin/Service/Model/LeaveListModal.php:45]: one field missing. termination id. this used to show whether the employee is past employee.
if this param null means active employee. if this param has value then we are showing `(Past Employee)` suffix

File: src/plugins/orangehrmLeavePlugin/config/routes.yaml [hunk 6]
```diff
@@ -206,6 +206,13 @@ apiv2_leave_reports_data:
   defaults:
     _api: OrangeHRM\Leave\Api\LeaveReportDataAPI
 
+apiv2_leave_leaves_list:
+  path: /api/v2/leave/leaves/list
```
Comment (RajithaKumara) [inline src/plugins/orangehrmLeavePlugin/config/routes.yaml:210]: will discuss and move to dashboard plugin

File: src/plugins/orangehrmLeavePlugin/test/Dao/LeaveListDaoTest.php [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+ * OrangeHRM is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with this program;
+ * if not, write to the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA  02110-1301, USA
+ */
+
+namespace Dao;
```
Comment (RajithaKumara) [inline src/plugins/orangehrmLeavePlugin/test/Dao/LeaveListDaoTest.php:20]:
```suggestion
namespace OrangeHRM\Tests\Leave\Dao;
```

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

File: src/plugins/orangehrmLeavePlugin/Service/Model/LeaveListModal.php [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+            ['employee','lastName'],
+            ['employee','firstName'],
+            ['employee','middleName'],
+            ['employee','employeeId'],
+            'duration',
+            'endTime',
+            'startTime',
+        ];
+
+        if ($this->getAuthUser()->getUserRoleName() === 'Admin') {
```
Comment (RajithaKumara) [inline src/plugins/orangehrmLeavePlugin/Service/Model/LeaveListModal.php:65]: TODO:: will change this to evaluate using datagroup permission

File: src/plugins/orangehrmDashboardPlugin/Api/EmployeeOnLeaveAPI.php [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+    use EmployeeOnLeaveServiceTrait;
+
+    public const DATE = 'date';
+
+    /**
+     * @inheritDoc
+     */
+    public function getAll(): EndpointResult
+    {
+        $leaveListSearchFilterParams = new LeaveListSearchFilterParams();
```
Comment (RajithaKumara) [inline src/plugins/orangehrmDashboardPlugin/Api/EmployeeOnLeaveAPI.php:48]:
```suggestion
        $employeeOnLeaveSearchFilterParams = new EmployeeOnLeaveSearchFilterParams();
```
:thinking:  can refactor rename

File: src/plugins/orangehrmDashboardPlugin/Api/EmployeeOnLeaveAPI.php [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+
+        $this->setSortingAndPaginationParams($leaveListSearchFilterParams);
+        $startDate = $this->getRequestParams()->getDateTime(
+            RequestParams::PARAM_TYPE_QUERY,
+            self::DATE,
+        );
```
Comment (RajithaKumara) [inline src/plugins/orangehrmDashboardPlugin/Api/EmployeeOnLeaveAPI.php:54]:
```suggestion
        $startDate = $this->getRequestParams()->getDateTime(
            RequestParams::PARAM_TYPE_QUERY,
            self::DATE,
            null,
            $this->getDateTimeHelper()->getNow()
        );
```
define default date.

`use DateTimeHelperTrait;` use this trait

File: src/plugins/orangehrmDashboardPlugin/Api/EmployeeOnLeaveAPI.php [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+
+        $leaveListSearchFilterParams->setDate($startDate);
```
Comment (RajithaKumara) [inline src/plugins/orangehrmDashboardPlugin/Api/EmployeeOnLeaveAPI.php:56]:
```suggestion
        $leaveListSearchFilterParams->setDate($date);
```
refactor rename

File: src/plugins/orangehrmDashboardPlugin/Api/EmployeeOnLeaveAPI.php [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+
+    /**
+     * @inheritDoc
+     */
+    public function getValidationRuleForGetAll(): ParamRuleCollection
+    {
+        return new ParamRuleCollection(
+            new ParamRule(
+                self::DATE,
+                new Rule(Rules::DATE)
```
Comment (RajithaKumara) [inline src/plugins/orangehrmDashboardPlugin/Api/EmployeeOnLeaveAPI.php:78]:
```suggestion
                new Rule(Rules::API_DATE)
```

File: src/plugins/orangehrmDashboardPlugin/Api/Model/EmployeeOnLeaveModal.php [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+use OrangeHRM\Core\Api\V2\Serializer\Normalizable;
+use OrangeHRM\Core\Traits\Auth\AuthUserTrait;
+use OrangeHRM\Core\Traits\Service\ConfigServiceTrait;
+use OrangeHRM\Core\Traits\UserRoleManagerTrait;
+use OrangeHRM\Entity\Leave;
+
+class EmployeeOnLeaveModal implements Normalizable
+{
+    use ModelTrait;
+    use ConfigServiceTrait;
```
Comment (RajithaKumara) [inline src/plugins/orangehrmDashboardPlugin/Api/Model/EmployeeOnLeaveModal.php:32]:
```suggestion
```

File: src/plugins/orangehrmDashboardPlugin/Api/Model/EmployeeOnLeaveModal.php [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+            ['employee','middleName'],
+            ['employee','employeeId'],
+            ['employee','terminationId'],
+            'duration',
+            'endTime',
+            'startTime',
+        ];
+
+        //TODO:: will change this to evaluate using data-group permission
+        if ($this->getUserRoleManager()->getUser()->getUserRole()->getName() === 'Admin') {
```
Comment (RajithaKumara) [inline src/plugins/orangehrmDashboardPlugin/Api/Model/EmployeeOnLeaveModal.php:69]:
```suggestion
        if ($this->getAuthUser()->getUserRoleName() === 'Admin') {
```
can you try this now

File: src/plugins/orangehrmDashboardPlugin/Dao/EmployeeOnLeaveDao.php [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with this program;
+ * if not, write to the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA  02110-1301, USA
+ */
+
+namespace OrangeHRM\Dashboard\Dao;
+
+use Doctrine\ORM\Tools\Pagination\Paginator;
```
Comment (RajithaKumara) [inline src/plugins/orangehrmDashboardPlugin/Dao/EmployeeOnLeaveDao.php:22]:
```suggestion
use OrangeHRM\ORM\Paginator;
```
careful about namespaces

File: src/plugins/orangehrmDashboardPlugin/Dao/EmployeeOnLeaveDao.php [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+    public function getEmployeeOnLeaveList(LeaveListSearchFilterParams $leaveListSearchFilterParams): array
+    {
+        return $this->getLeaveListPaginator($leaveListSearchFilterParams)->getQuery()->execute();
+    }
+
+    /**
+     * @param LeaveListSearchFilterParams $leaveListSearchFilterParams
+     * @return Paginator
+     */
+    public function getLeaveListPaginator(LeaveListSearchFilterParams $leaveListSearchFilterParams): Paginator
```
Comment (RajithaKumara) [inline src/plugins/orangehrmDashboardPlugin/Dao/EmployeeOnLeaveDao.php:45]:
```suggestion
    private function getLeaveListPaginator(LeaveListSearchFilterParams $leaveListSearchFilterParams): Paginator
```

File: src/plugins/orangehrmDashboardPlugin/test/Api/EmployeeOnLeaveApiTest.php [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+ * OrangeHRM is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with this program;
+ * if not, write to the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA  02110-1301, USA
+ */
+
+namespace OrangeHRM\Dashboard\test\Api;
```
Comment (RajithaKumara) [inline src/plugins/orangehrmDashboardPlugin/test/Api/EmployeeOnLeaveApiTest.php:20]:
```suggestion
namespace OrangeHRM\Tests\Dashboard\Api;
```

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

File: src/plugins/orangehrmDashboardPlugin/Api/Model/EmployeeOnLeaveModal.php (hunk 14 above)
Comment (ManulMax) [reply src/plugins/orangehrmDashboardPlugin/Api/Model/EmployeeOnLeaveModal.php]: working

File: src/plugins/orangehrmDashboardPlugin/Api/Model/EmployeeOnLeaveModal.php [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+            'startTime',
+        ];
+
+        //TODO:: will change this to evaluate using data-group permission
+        if ($this->getAuthUser()->getUserRoleName() === 'Admin') {
+            $filter[] = ['getLeaveType', 'getId'];
+            $filter[] = ['getLeaveType', 'getName'];
+
+            $attributeNames[] = ['leaveType', 'id'];
+            $attributeNames[] = ['leaveType', 'name'];
```
Comment (RajithaKumara) [inline src/plugins/orangehrmDashboardPlugin/Api/Model/EmployeeOnLeaveModal.php:73]: Oops leave type deleted attribute forgot. https://github.com/orangehrm/orangehrm/blob/main/src/plugins/orangehrmLeavePlugin/Api/Model/LeaveTypeModel.php#L40

File: src/plugins/orangehrmDashboardPlugin/test/fixtures/testcases/EmployeeOnLeaveAPITestCases.yaml [hunk 19]
```diff
@@ -0,0 +1,115 @@
+GetAll:
```
Comment (RajithaKumara) [inline src/plugins/orangehrmDashboardPlugin/test/fixtures/testcases/EmployeeOnLeaveAPITestCases.yaml:1]: seems not added tests for `limit`, `offset`, `sortOrder`, `sortField`

**Review by RajithaKumara — APPROVED** [review]

**Pull request merged**